### DISCUSSION TITLE
POA based mismatch

### DISCUSSION THREAD

**Discussion by kurt-rhee** [pvlib/pvlib-python#1505] [General]
Hello NREL, 

I was wondering how one would model the effects of differences in POA on the combined IV curve controlled by a single MPPT.  For example, if we had 2 horizontal single axis trackers back-tracking independently of one another, but attached to the same MPPT.  In the example HSAT 1 could have a rotation angle of -55 degrees and HSAT could have a rotation angle of -50 degrees.  For simplicity's sake, if HSAT 1 had one string and HSAT 2 had another string, how would their IV curves be combined?

Thank you for the help.

**Comment by cwhanse** [comment]
Hi Kurt,

Are you considering the tracking arrays to be conneced in series or in parallel?

If in series: there are functions in `pvlib.singlediode` which can compute each array's IV curve, but you'll have to do the "addition" (combining one array's curve with the other) with your own code. pvlib-python currently has no functions that model MPPT algorithms or the internal voltage controls of an inverter, so you will also have to locate the maximum power point on the combined IV curve.

If in parallel: the two arrays should be at similar temperature so the DC voltages at each array's maximum power point may not be that different. If that's the case, I would average the DC voltages, add the DC power, and use one of the pvlib inverter models to convert DC to AC.

**Reply by kandersolar** [reply]
If you do find it necessary to combine complete IV curves, here's an example showing one (clunky) way of going about it: https://pvlib-python.readthedocs.io/en/stable/gallery/shading/plot_partial_module_shading_simple.html

That example is for calculating module IV curves under different cell-level operation conditions, but the method should apply equally well to array IV curves under different string-level operating conditions.

**Comment by kandersolar** [comment]
> Hello NREL,

P.S. pvlib is a community project supported by contributors from many organizations across the national labs, academia, and industry.  It's not overseen or led by any one institution, least of all NREL.  In fact, considering the full history of the project, NREL's involvement began relatively recently and is pretty minor with only one frequent contributor (me).  Just pointing it out so that everyone here gets their due credit :)

**Reply by kurt-rhee** [reply]
Hey 
@kanderso-nrel I'm currently writing a paper that references some of your work in terms of modeling backtracking strategies on terrain.  I was wondering if you would like to give it a read and send feedback before I submit it for publication.

**Reply by kandersolar** [reply]
Sure, I'd be happy to!  My email is kevin.anderson@nrel.gov.  @mikofski and co might be interested too ("Tracker Terrain Losses" [Part 1](https://ieeexplore.ieee.org/abstract/document/9300381) and [Part 2](https://ieeexplore.ieee.org/abstract/document/9573469)).

**Reply by kurt-rhee** [reply]
Awesome, I'll send a draft today.  Haha I agree that @mikofski  might be interested.  I sent an email to him separately earlier.  Mark if you are reading this I was able to download copies of both papers from IEEE.

**Reply by mikofski** [reply]
@kurt-rhee oh cool, sorry I didn't get back to you faster. Glad you got copies, and yes, I'd ❤️ to read your draft if you're okay to share. Thanks!

**Comment by kurt-rhee** [comment]
Hey Cliff, 

Sorry for addressing this to NREL and not to everyone!  Thanks @kanderso-nrel  for the correction.

I was thinking that they would be connected in parallel.  Your answer makes total sense, thank you very much.  

I was also wondering what would happen if one of the paralleled strings were shaded or partially shaded?  In the case of total shading my thought is the unshaded string would equal 100% of the MPP contribution.  In the case of the partially shaded string (lets say bottom 1/3 of all modules in the string are shaded, cells are not half-cut), my thought is that the unshaded string would also by 100% of the MPP contribution.  

Am I thinking about this the right way?

**Comment by kurt-rhee** [comment]
Actually, I found this note from solaredge which describes multiple string installations on page 3.  If I'm reading correctly it seems like in the partially shaded case, the percentage of shading determines how much power is produced.

https://www.solaredge.com/sites/default/files/se_technical_bypass_diode_effect_in_shading.pdf

**Reply by cwhanse** [reply]
That is true when micro-inverters are used "A SolarEdge system, with module-level MPP trackers..."  For central or string inverters, that is not true, in general. The power is reduced roughly in proportion to the number of cell-strings affected by the shading.

**Reply by mikofski** [reply]
>I was also wondering what would happen if one of the paralleled strings were shaded or partially shaded?

* If the shaded string had uniform shade, _IE_ reduced irradiance over all cells in all modules of the string, then this would reduce the current, but not the voltage and would be a similar situation to what Cliff described above: "Average the voltages and add the powers"
* If the shaded string has any _non-uniform_ shade, then it might depend on a couple of factors:
    * if the shade affects all of the modules in the string, _IE_ a shade line that either runs across the bottom fraction of all the modules of the string, and the modules are **not** halfcut then the rule of thumb is that the string will typically perform as well as the _worst_ cell in the string
    * if the shade only affects a few substrings of modules in a string, then it may depend on how large the array is, how many shaded and unshaded strings there are, and how badly shaded those substrings are.
        * In a large array with very few shaded substrings, it's possible that the the bypass diodes in shaded substrings will not be activated because the power gained from bypassing those substrings is less than the power lost by operating at a lower voltage.
        * However in a 2 string array or an array with more shaded strings than unshaded, it's possible that the bypass diodes of the shaded substrings will activate, dropping the voltage for the entire array by the number of substrings. That means that you would need to calculate the power for the unshaded string at the new voltage and add it to the power from the unshaded substrings. This gets complicated quickly if the shading differs significantly between strings, and I would suggest using PVMismatch.

> In the case of total shading my thought is the unshaded string would equal 100% of the MPP contribution. 

I think this is an approximation, b/c there would still be diffuse light on the shaded string so it could still contribute about 10-20% of its power.

> In the case of the partially shaded string (lets say bottom 1/3 of all modules in the string are shaded, cells are not half-cut), my thought is that the unshaded string would also by 100% of the MPP contribution.

I agree, this scenario is similar to the "total shading", it should perform as well as the worst cell in the string, but the shaded third of the modules would still have incident diffuse irradiance, so could still contribute 10-20% of its power.

## PVMismatch sample
The no shade, total shade, and third shade scenarios yield the following:
```python
>>> pvsys.Pmp
# 6425.606754448506 [W]

>>> pvsys_total_shade.Pmp
# 3447.351878214916 [W]

>>> pvsys_third_shade.Pmp
# 3535.148118237753 [W]
```
So the total shade and third shade losses are about the same, almost 50% loss, but actually a little less than half because I left 10% diffuse light on the "shaded" cells.

<details>

```python
# import the pvmismatch package
from pvmismatch import *
from matplotlib import pyplot as plt
plt.ion()  # make ploting interactive

# set the resolution of the IV curves to about 3000 points
pvconst = pvconstants.PVconstants(npts=1001)

# create the unshaded system with 2 strings of 10 modules per string
# by default this uses SPR-E20-315, a 315-W SunPower Series-E module
pvsys = pvsystem.PVsystem(pvconst=pvconst, numberStrs=2, numberMods=10)
pvsys.plotSys()  # plot the system
plt.tight_layout()  # clean up the subplots
pvsys.pvstrs[0].plotStr()  # plot the 1st substring
plt.tight_layout()

# display power
pvsys.Pmp

# create an identical system to simulate total shading of direct beam irrad
pvsys_total_shade = pvsystem.PVsystem(pvconst=pvconst, numberStrs=2, numberMods=10)

# shade the first substring (index 0) so only 10% of incidenct irrad (diffuse only)
pvsys_total_shade.setSuns({0: 0.1})
pvsys_total_shade.plotSys()  # plot system
plt.tight_layout()
fig = pvsys_total_shade.pvstrs[0].plotStr()  # plot shaded substring
plt.tight_layout()
fig.axes[0].set_ylim([0, 6])  # limit current plot to 6-A
fig.axes[1].set_ylim([0, 300])  # limit power plot to 300-W

# display power
pvsys_total_shade.Pmp

# create another identical system for 1/3 shade scenario
pvsys_third_shade = pvsystem.PVsystem(pvconst=pvconst, numberStrs=2, numberMods=10)
# indices of shaded cells in bottom third of module
# note: default module, SPR-E20-315, has 96 cells in 8 columns in serpentine pattern
pvmod_shade_cells = [
     8,  9, 10, 11, 12, 13, 14, 15,
    32, 33, 34, 35, 36, 37, 38, 39,
    56, 57, 58, 59, 60, 61, 62, 63,
    80, 81, 82, 83, 84, 85, 86, 87]
# apply shading to all modules in first substring
pvsys_third_shade.setSuns({0: {pvmod_id: {'cells': pvmod_shade_cells, 'Ee': 0.1} for pvmod_id in range(10)}})
pvsys_third_shade.plotSys()  # plot
plt.tight_layout()
pvsys_third_shade.Pmp
```

</details>

### no shade PV system
![noshade_pvsys](https://user-images.githubusercontent.com/1385621/181074692-4137b66c-adef-47bc-bcd1-8f10eaf636b3.png)

### total shade PV system
![totshade_pvsys](https://user-images.githubusercontent.com/1385621/181074845-026a5b87-f51e-4c2b-afdf-b1b89d8c21ac.png)

### third shade PV system
![third_shade_pvsys](https://user-images.githubusercontent.com/1385621/181074903-6cfa6804-a0f4-4fd5-aea4-3311a3631252.png)

**Comment by mikofski** [comment]
@kurt-rhee  you might check out [PVMismatch](https://sunpower.github.io/PVMismatch/) it makes connecting different modules and strings with different irradiance and shading patterns easy. There’s even an app in the contrive folder by @bmeyers

**Comment by kurt-rhee** [comment]
@mikofski As always this is incredibly helpful, and I appreciate the guidance you've given me so much.  PVMismatch is incredible, I am so glad that it exists.  

Also thank you for all of the code and visual examples, it really makes the issue clear for me.
